core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -131,6 +139,16 @@ public void onComplete() {
                     }
                     // Releasing the lock to move ahead with the next request in queue.
                     releasePartitionLocks(shareFetchData.groupId(), topicPartitionData.keySet());
+                    // If we have a fetch request completed for a topic-partition, we release the locks for that partition,
+                    // then we should check if there is a pending share fetch request for the topic-partition and complete it.
+                    // We add the action to delayed actions queue to avoid an infinite call stack, which could happen if
+                    // we directly call delayedShareFetchPurgatory.checkAndComplete
+                    delayedActionQueue.add(() -> {
+                        result.keySet().forEach(topicIdPartition ->
+                            delayedShareFetchPurgatory.checkAndComplete(
+                                new DelayedShareFetchKey(shareFetchData.groupId(), topicIdPartition)));
+                        return BoxedUnit.UNIT;
+                    });
```

Comment:
What do you think about moving this logic to SharePartitionManager as discussed in https://github.com/apache/kafka/pull/17177#issuecomment-2392108397?

Reply:
hi @junrao @mumrah , yes, I think we can move it outside DelayedShareFetch class to SharePartitionManager. Also, I'll also add TopicPartition as a key for delayed share fetch along with SharePartition (that is already present right now). I have compiled these details in JIRA  https://issues.apache.org/jira/browse/KAFKA-17703. Will it be fine if we can merge this PR and I'll take up that JIRA in a future PR?